man/vde_switch.1
.TH VDE_SWITCH 1 "December 6, 2006" "Virtual Distributed Ethernet"
.SH NAME
vde_switch \- Virtual Distributed Ethernet switch
.SH SYNOPSIS
.B vde_switch 
[ 
.B \-hub  
] 
[ 
.BI \-sock 
.I commdirpath  
]  
[ 
.BI \-mod 
.I octal-mode  
]  
[ 
.BI \-group
.I NAME
]  
[ 
.BI \-tap 
.I interface  
] 
[
.BI \-daemon 
]
.br
.SH DESCRIPTION
A
\fBvde_switch\fP 
is a virtual switch for the vde architecture.
.br
A vde network can include several vde_switches running on different
(real) computers.
vde_switches can be connected by vde cables (see
\fBvde_plug (1)\fP
).
.br
Supported architecture for VDE end nodes include: 
real linux boxes (through the tuntap interface),
bochs, qemu and MPS virtual machines.
.br
VDE is useful to create networks of heterogeneous virtual machines as
well as as a general tunneling tool -- all the ethernet based protocols work
by this kind of tunnel -- and also as a tool for mobility.
In fact VDE creates an overlay network where users can keep their IP
addresses despite of the change of IP addresses on the interfaces.

vde_switch needs root privileges to open a tap interface, can be run by
users when no -tap option is specified.

vde-2 has several features: VLAN, Fast Spanning Tree Protocol, command line management
(on console for foreground switches, on a terminal for daemonized switches).
If a vde_switch run in foreground mode simply type return to have the prompt.
For daemons there is the -M option to specify the management socket. Unixterm is the tool
to start a terminal for the management.
Type "help" at the prompt for a list of possible options.

.SH OPTIONS
.TP
.B \-hub 
turn off the switch engine. It operates as a hub.
.TP
.B \-sock "\fIcommdirpath\fP"
specify the directory when comm socket are created.
The default value is "/tmp/vde.ctl"
.TP
.B \-mod "\fIoctalmode\fP"
specify the octal permissions for the comm sockets.
.TP
.B \-group "\fINAME\fP"
specify the group owner for the comm socket.
The default value is the current user's primary group
.TP
.B \-tap "\fIinterface\fP" 
connect the specified tuntap interface to this vde_switch (needs root privileges)
It is possible to specify several tap interfaces, comma separated.
.TP
.B \-daemon
Run as a daemon. Use syslog for logging.
.TP
.B \-f rcfile
rc file to set the initial switch configuration. This rc file has the same syntax of
the command line interface.
.TP
.B \-nostdin
Allow stdin to be closed.
.TP
.B \-h 
help. Please use it for a more complete list of options.
vde_switch is a modular program, options vary depending of the compiled-in modules.
.SH NOTICE
Virtual Distributed Ethernet is not related in any way with
www.vde.com ("Verband der Elektrotechnik, Elektronik und Informationstechnik"
i.e. the German "Association for Electrical, Electronic & Information
Technologies").

.SH SEE ALSO
\fBvde_plug\fP(1),
\fBvde_plug2tap\fP(1),
\fBdpipe\fP(1),
.br
.SH AUTHOR
VDE is a project by Renzo Davoli <renzo@cs.unibo.it>
.br
VDE started as an enhancement of uml_switch by Jeff Dike and others.
VDE-2 has been almost completely rewritten but maybe some statements
come from the historical source. So, some copyright and thanks also for Dike&Co.


Changelog
VDE 2.3.3, 2022-05-02
	* Several bug fixes
	* new hash function
	* Allow pipes on stdin for daemonized switches
	* fixed wirefilter '-N' option
	* vde_switch: min packet len enforced
	* Added vde_router
	* dpipe: Fixed error checking for setpgrp().
	* Fix incorrect chacha key size
	* configure: do not test for CXX, it's not needed
	* Fix description for "vlan/delport" command.
	* Silence compiler warning about implicit type.
	* Fix typo in error message
	* Fixes 2019.10.06
	* fstp: Add static to inline functions
	* configure: Check for wolfSSL instead of OpenSSL.
	* build: Fix parallel make.
	* vde_switch: add `-nostdin` flag to allow stdin to be closed
	* Deprecate modules
	* Fix warnings with newer gcc
	* [vde_plug] Removed kvde IPN support
	* In vde_switch retry send() when it returns ENOBUFS on Darwin and FreeBSD
	* configure.ac: Downgrade autoconf from 2.71 to 2.59
	* Added GitHub actions

VDE 2.3.2, 2011-11-23
	* dpipe: new daemon mode
	* [vde_switch] new mainloop management
	* [vde_switch] hash table based on 64bits integers
	* [vde_switch] new priority queue for packets
	* wirefilter: capacity renamed as "channel buffer size"

VDE 2.3.1, 2010-07-25
	* [vde_switch] Fix control socket path resolution regression.

VDE 2.3.0, 2010-05-29
	* Remove some hardcoded paths and use SYSCONFDIR/LOCALSTATEDIR instead.
	* [vde_switch] Fix multiple vlan tagging when packetfilter is used.
	* Build system fixes.
	* Ship README and patch to add VDE support in VirtualBox (included in
	  VirtualBox since 3.2.0).
	* Ship an updated version of slirp taken from QEMU and include VDE support.
	* [vde_cryptcab] Remove SIGALRM-based timers.
	* [vde_l3] Fix unicast test (sourceforge: #2726725)
	* [libvdeplug] Add vdestream abstraction to send/receive bytes (previously in
	  vde_plug)
	* [vde_switch] New command "port/createauto" to create a new port with an
	  automatically allocated ID

VDE 2.2.3, 2009-05-05
	* [vdeterm] Moved history management in a separate library.
	* [vde_switch] Console bug fixes.
	* Added vde_pcapplug (connect a vde_switch to a real interface using pcap)
	* Added missing includes of limits.h/ctype.h
	* [vde_switch] Added option --mgmtgroup to consmgmt module to specify the
	  group of the management socket.
	* [vde_switch] Added option --dirmode to datasock module to specify the mode
	  of the socket directory.
	* [vdetaplib] Added specification of port, mode and group.
	* [kvde_switch] Bugfixes.
	* [vde_switch] pdump plugin: add fifo support and buffered/unbuffered choice.
	* [common] Fix '#include <limits.h>' for PATH_MAX. (Closes: #2023790)
	* [vdetaplib] Manpage: added an example to override default port, group and
	  mode. (Closes: #2066885)
	* [vde_switch] Resize untagged ports bitarray while changing number of ports
	  (Thanks to Michael Sallaway). Closes: #2115547.
	* [vde_switch] Fix memory reset in BA_REALLOC and BAC_REALLOC (by rd235).
	  (Closes: #2123601)
	* [kvde_switch] By default kvde_switch is now not compiled in.
	* [vdeq] Fixed loop if an unknow parameter is specified. Thanks to tfero.
	  (Closes: #2134438)
	* [slirpvde] Imported patch from <ludwig.nussel@suse.de> to fix static buffer
	  overflow (Closes: #2138410).
	* [vde_switch] Fixed compatibility with armel architecture using "attribute
	  packet" on registration message structures (thanks to Michele Cucchi).
	* [vde_cryptcap] Some modifications to allow the client to work on the
	  OpenMoko freerunner.
	* [vde_switch] Added port access control.
	* [vde_switch] Changed regular user standard switch to ~/.vde2/stdsock.
	* [libvdeplug] Downsize vde_open_real() and vde_realpath() stack to fit in
	  User-Mode Linux.
	* [common] Add pkg-config files for VDE libraries.
	* [common] Add '--disable-pcap' configure option, thanks to Wulf C. Krueger
	  (Closes: #2600817).
	* [common] Fix bad malloc replacement (Closes: #2631581), thanks to Michele
	  Cucchi.
	* [vde_cryptcab] Added support for daemonization.
	* [vde_cryptcab] Fixed server side disconnections (thanks to Brendan Grieve).
	* [wirefilter] Added rc file.

VDE 2.2.2, 2008-07-08
	* Added --enable-profile to configure (for profiling)
	* Added cleanup() in vdeq on successfull termination (thanks huslu)
	* Added --mod to new syntax in vdeq man page
	* Fixed fd leak in runscript() (Closes: #2009311 on SF)
	* wirefilter man and code cleanup
	* wirefilter option for blinking added.
	* Gaussian distribution for range values, Gilbert bursty model for packet loss.
	* Bugfix on mgmt sessions.
	* Poor's man hub implemented. (patch by Luca Saiu)

VDE 2.2.1, 2008-06-17
	* New contact email address: <info@v2.cs.unibo.it>
	* Code cleanup (some more warnings removed)
	* [vde_crytcab] Cleared large memory leak due to missing
	  EVP_CIPHER_CTX_cleanup() after encryption/decryption
	* [vde_cryptcab] Cleared (very small) memory leak in crc32 calls
	* [vde_cryptcab] Fixed bug on handover after session timeout: now client
	  re-establishes the connection when it sees some traffic, after session
	  timeout.
	* Heavy directories reorganization, all source code moved under src and
	  include
	* New internal library libvdecommon.la with common utilities and LIBOBJ
	  replacement/compatibility functions
	* [common] Fix in poll emulation management
	* Fix to distclean problem due to wrong LIBOBJ management
	* [vde_switch] hash/setsize 0 was not recognized as an error and caused vde_switch
	  to hang due to an infinite loop
	* Don't compile cryptcab if configure detects not to.
	* Fix for autotools in Slackware
	* [vde_cryptcab] refactoring:
	  - simplified protocol (though keeping it back-compatible)
	  - fixed some bugs and memory leaks
	  - added some new features
	* [vde_switch] fix on BPDU generation/handling: do not generate BPDU if FSTP is not active
	  or we are HUB, forward BPDU if FSTP is not active (Closes #1943973 on SF).
	* [slirpvde] include limits.h on linux as well.
	* [vde_switch] Applied patch from Bjorn Mork to remove 32/64 bit confusion from bitarray.h

vde-2/vde_switch.c
/* Copyright 2005 Renzo Davoli VDE-2
 * Licensed under the GPL
 * --pidfile/-p and cleanup management by Mattia Belletti.
 * some code remains from uml_switch Copyright 2001, 2002 Jeff Dike and others
 * Modified by Ludovico Gardenghi 2005
 */

#include <unistd.h>
#define _GNU_SOURCE
#include <getopt.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <signal.h>
#include <syslog.h>
#include <errno.h>
#ifndef HAVE_BROKEN_POLL
#include <sys/poll.h>
#endif
#include <switch.h>
#include <config.h>
#include <qtimer.h>
#include <hash.h>
#include <port.h>
#ifdef FSTP
#include <fstp.h>
#endif
#include <consmgmt.h>
#include <sys/time.h>
#include <time.h>

#include <vde.h>
#ifdef VDE_PQ
#include <packetq.h>
#endif

#ifdef HAVE_BROKEN_POLL
#include "poll2select.h"
#define poll poll2select
#endif

static struct swmodule *swmh;

char *prog;
unsigned char switchmac[ETH_ALEN];
unsigned int priority=DEFAULT_PRIORITY;

static int hash_size=INIT_HASH_SIZE;
static int numports=INIT_NUMPORTS;


static void recaddswm(struct swmodule **p,struct swmodule *new)
{
	struct swmodule *this=*p;
	if (this == NULL)
		*p=new;
	else 
		recaddswm(&(this->next),new);
}

void add_swm(struct swmodule *new)
{
	static int lastlwmtag;
	new->swmtag= ++lastlwmtag;
	if (new != NULL && new->swmtag != 0) {
		new->next=NULL;
		recaddswm(&swmh,new);
	}
}

static void recdelswm(struct swmodule **p,struct swmodule *old)
{
	struct swmodule *this=*p;
	if (this != NULL) {
		if(this == old)
			*p=this->next;
		else
			recdelswm(&(this->next),old);
	}
}

void del_swm(struct swmodule *old)
{
	if (old != NULL) {
		recdelswm(&swmh,old);
	}
}

/* FD MGMT */
struct pollplus {
	unsigned char type;
	int arg;
	time_t timestamp;
};

static int nfds = 0;
static int nprio =0;
static struct pollfd *fds = NULL;
static struct pollplus **fdpp = NULL;

static int maxfds = 0;

static struct swmodule **fdtypes;
static int ntypes;
static int maxtypes;

#define PRIOFLAG 0x80
#define TYPEMASK 0x7f
#define ISPRIO(X) ((X) & PRIOFLAG)

#define TYPE2MGR(X) (fdtypes[((X) & TYPEMASK)])

unsigned char add_type(struct swmodule *mgr,int prio)
{
	register int i;
	if(ntypes==maxtypes) {
		maxtypes = maxtypes ? 2 * maxtypes : 8;
		if (maxtypes > PRIOFLAG) {
			printlog(LOG_ERR,"too many file types");
			exit(1);
		}
		if((fdtypes = realloc(fdtypes, maxtypes * sizeof(struct swmodule *))) == NULL){
			printlog(LOG_ERR,"realloc fdtypes %s",strerror(errno));
			exit(1);
		}
		memset(fdtypes+ntypes,0,sizeof(struct swmodule *) * maxtypes-ntypes);
		i=ntypes;
	} else
		for(i=0; fdtypes[i] != NULL; i++)
			;
	fdtypes[i]=mgr;
	ntypes++;
	return i | ((prio != 0)?PRIOFLAG:0);
}

void del_type(unsigned char type)
{
	type &= TYPEMASK;
	if (type < maxtypes)
		fdtypes[type]=NULL;
	ntypes--;
}

void add_fd(int fd,unsigned char type,int arg)
{
	struct pollfd *p;
	int index;
	/* enlarge fds and g_fdsdata array if needed */
	if(nfds == maxfds){
		maxfds = maxfds ? 2 * maxfds : 8;
		if((fds = realloc(fds, maxfds * sizeof(struct pollfd))) == NULL){
			printlog(LOG_ERR,"realloc fds %s",strerror(errno));
			exit(1);
		}
		if((fdpp = realloc(fdpp, maxfds * sizeof(struct pollplus *))) == NULL){
			printlog(LOG_ERR,"realloc pollplus %s",strerror(errno));
			exit(1);
		}
	}
	if (ISPRIO(type)) {
		fds[nfds]=fds[nprio];
		fdpp[nfds]=fdpp[nprio];
		index=nprio;
		nprio++;
	} else
		index=nfds;
	if((fdpp[index]=malloc(sizeof(struct pollplus))) == NULL) {
		printlog(LOG_ERR,"realloc pollplus elem %s",strerror(errno));
		exit(1);
	}
	p = &fds[index];
	p->fd = fd;
	p->events = POLLIN | POLLHUP;
	fdpp[index]->type=type;
	fdpp[index]->arg=arg;
	nfds++;
}

static void file_cleanup(void)
{
	register int i;
	for(i = 0; i < nfds; i++)
		TYPE2MGR(fdpp[i]->type)->cleanup(fdpp[i]->type,fds[i].fd,fdpp[i]->arg);
}

void remove_fd(int fd)
{
	register int i;

	for(i = 0; i < nfds; i++){
		if(fds[i].fd == fd) break;
	}
	if(i == nfds){
		printlog(LOG_WARNING,"remove_fd : Couldn't find descriptor %d", fd);
	} else {
		struct pollplus *old=fdpp[i];
		TYPE2MGR(fdpp[i]->type)->cleanup(fdpp[i]->type,fds[i].fd,fdpp[i]->arg);
		if (ISPRIO(fdpp[i]->type)) nprio--;
		memmove(&fds[i], &fds[i + 1], (maxfds - i - 1) * sizeof(struct pollfd));
		memmove(&fdpp[i], &fdpp[i + 1], (maxfds - i - 1) * sizeof(struct pollplus *));
		free(old);
		nfds--;
	}
}


static void main_loop()
{
	time_t now;
	register int n,i;
	while(1) {
#ifdef VDE_PQ
		n=poll(fds,nfds,packetq_timeout);
#else
		n=poll(fds,nfds,-1);
#endif
		now=qtime();
		if(n < 0){ 
			if(errno != EINTR)
				printlog(LOG_WARNING,"poll %s",strerror(errno));
		} else {
			for(i = 0; /*i < nfds &&*/ n>0; i++){
				if(fds[i].revents != 0) {
					register int prenfds=nfds;
					n--;
					fdpp[i]->timestamp=now;
					TYPE2MGR(fdpp[i]->type)->handle_input(fdpp[i]->type,fds[i].fd,fds[i].revents,&(fdpp[i]->arg));
					if (nfds!=prenfds) /* the current fd has been deleted */
						break; /* PERFORMANCE it is faster returning to poll */
				}	
/* optimization: most used descriptors migrate to the head of the poll array */
#ifdef OPTPOLL
				else
				{
					if (i < nfds && i > 0 && i != nprio) {
						register int i_1=i-1;
						if (fdpp[i]->timestamp > fdpp[i_1]->timestamp) {
							struct pollfd tfds;
							struct pollplus *tfdpp;
							tfds=fds[i];fds[i]=fds[i_1];fds[i_1]=tfds;
							tfdpp=fdpp[i];fdpp[i]=fdpp[i_1];fdpp[i_1]=tfdpp;
						}
					}
				}
#endif
			}
#ifdef VDE_PQ
			if (packetq_timeout > 0)
				packetq_try();
#endif
		}
	}
}

/* starting/ending routines, main_loop, main*/
#define HASH_TABLE_SIZE_ARG 0x100
#define MACADDR_ARG         0x101
#define PRIORITY_ARG        0x102

static void Usage(void) {
	struct swmodule *p;
	printf(
			"Usage: vde_switch [OPTIONS]\n"
			"Runs a VDE switch.\n"
			"(global opts)\n"
			"  -h, --help                 Display this help and exit\n"
			"  -v, --version              Display informations on version and exit\n"
			"  -n  --numports             Number of ports (default %d)\n"
			"  -x, --hub                  Make the switch act as a hub\n"
#ifdef FSTP
			"  -F, --fstp                 Activate the fast spanning tree protocol\n"
#endif
			"      --macaddr MAC          Set the Switch MAC address\n"
#ifdef FSTP
			"      --priority N           Set the priority for FST (MAC extension)\n"
#endif
			"      --hashsize N           Hash table size\n"
			,numports);
	for(p=swmh;p != NULL;p=p->next)
		if (p->usage != NULL)
			p->usage();
	printf(
			"\n"
			"Report bugs to "PACKAGE_BUGREPORT "\n"
			);
	exit(1);
}

static void version(void)
{ 
	printf(
			"VDE " PACKAGE_VERSION "\n"
			"Copyright 2003,2004,2005 Renzo Davoli\n"
			"some code from uml_switch Copyright (C) 2001, 2002 Jeff Dike and others\n"
			"VDE comes with NO WARRANTY, to the extent permitted by law.\n"
			"You may redistribute copies of VDE under the terms of the\n"
			"GNU General Public License v2.\n"
			"For more information about these matters, see the files\n"
			"named COPYING.\n");
	exit(0);
} 

static struct option *optcpy(struct option *tgt, struct option *src, int n, int tag)
{
	register int i;
	memcpy(tgt,src,sizeof(struct option) * n);
	for (i=0;i<n;i++) {
		tgt[i].val=(tgt[i].val & 0xffff) | tag << 16;
	}
	return tgt+n;
}

static int parse_globopt(int c, char *optarg)
{
	int outc=0;
	switch (c) {
		case 'x':
			portflag(P_SETFLAG,HUB_TAG);
			break;
		case HASH_TABLE_SIZE_ARG:
			sscanf(optarg,"%i",&hash_size);
			break;
#ifdef FSTP
		case 'F':
			fstflag(P_SETFLAG,FSTP_TAG);
			break;
		case PRIORITY_ARG:
			sscanf(optarg,"%i",&priority);
			priority &= 0xffff;
			break;
#endif
		case MACADDR_ARG:
			{int maci[ETH_ALEN],rv;
				if (index(optarg,':') != NULL)
					rv=sscanf(optarg,"%x:%x:%x:%x:%x:%x", maci+0, maci+1, maci+2, maci+3, maci+4, maci+5);
				else
					rv=sscanf(optarg,"%x.%x.%x.%x.%x.%x", maci+0, maci+1, maci+2, maci+3, maci+4, maci+5);
				if (rv != 6) {
					printlog(LOG_ERR,"Invalid MAC Addr %s",optarg);
					Usage();
				}
				else  {
					register int i;
					for (i=0;i<ETH_ALEN;i++)
						switchmac[i]=maci[i];
				}
			}
			break;
		case 'n':
			sscanf(optarg,"%i",&numports);
			break;
		case 'v':
			version();
			break;
		case 'h':
			Usage();
			break;
		default:
			outc=c;
	}
	return outc;
}

static void parse_args(int argc, char **argv)
{
	struct swmodule *swmp;
	struct option *long_options;
	char *optstring;
	static struct option global_options[] = {
		{"help",0 , 0, 'h'},
		{"hub", 0, 0, 'x'},
#ifdef FSTP
		{"fstp",0 , 0, 'F'},
#endif
		{"version", 0, 0, 'v'},
		{"numports", 1, 0, 'n'},
		{"hashsize", 1, 0, HASH_TABLE_SIZE_ARG},
		{"macaddr", 1, 0, MACADDR_ARG},
#ifdef FSTP
		{"priority", 1, 0, PRIORITY_ARG}
#endif
	};
	static struct option optail = {0,0,0,0};
#define N_global_options (sizeof(global_options)/sizeof(struct option))
	prog = argv[0];
	int totopts=N_global_options+1;

	for(swmp=swmh;swmp != NULL;swmp=swmp->next)
		totopts += swmp->swmnopts;
	long_options=malloc(totopts * sizeof(struct option));
	optstring=malloc(2 * totopts * sizeof(char));
	if (long_options == NULL || optstring==NULL)
		exit(2);
	{ /* fill-in the long_options fields */
		register int i;
		char *os=optstring;
		char last=0;
		struct option *opp=long_options;
		opp=optcpy(opp,global_options,N_global_options,0);
		for(swmp=swmh;swmp != NULL;swmp=swmp->next)
			opp=optcpy(opp,swmp->swmopts,swmp->swmnopts,swmp->swmtag);
		optcpy(opp,&optail,1,0);
		for (i=0;i<totopts-1;i++)
		{
			int val=long_options[i].val & 0xffff;
			if(val > ' ' && val <= '~' && val != last)
			{
				*os++=val;
				if(long_options[i].has_arg) *os++=':';
			}
		}
		*os=0;
	}
	{
		/* Parse args */
		int option_index = 0;
		int c;
		while (1) {
			c = GETOPT_LONG (argc, argv, optstring,
					long_options, &option_index);
			if (c == -1)
				break;
			c=parse_globopt(c,optarg);
			for(swmp=swmh;swmp != NULL && c!=0;swmp=swmp->next) {
				if (swmp->parseopt != NULL) {
					if((c >> 7) == 0)
						c=swmp->parseopt(c,optarg);
					else if ((c >> 16) == swmp->swmtag)
						swmp->parseopt(c & 0xffff,optarg),c=0;
				}
			}
		}
	}
	if(optind < argc)
		Usage();
	free(long_options);
	free(optstring);
}

static void init_mods(void)
{
	struct swmodule *swmp;
	for(swmp=swmh;swmp != NULL;swmp=swmp->next)
		if (swmp->init != NULL)
			swmp->init();
}

static void cleanup(void)
{
	struct swmodule *swmp;
	file_cleanup();
	for(swmp=swmh;swmp != NULL;swmp=swmp->next)
		if (swmp->cleanup != NULL)
			swmp->cleanup(0,-1,-1);
}

static void sig_handler(int sig)
{
	printlog(LOG_ERR,"Caught signal %d, cleaning up and exiting", sig);
	cleanup();
	signal(sig, SIG_DFL);
	kill(getpid(), sig);
}

static void setsighandlers()
{
	/* setting signal handlers.
	 *      * sets clean termination for SIGHUP, SIGINT and SIGTERM, and simply
	 *           * ignores all the others signals which could cause termination. */
	struct { int sig; const char *name; int ignore; } signals[] = {
		{ SIGHUP, "SIGHUP", 0 },
		{ SIGINT, "SIGINT", 0 },
		{ SIGPIPE, "SIGPIPE", 1 },
		{ SIGALRM, "SIGALRM", 1 },
		{ SIGTERM, "SIGTERM", 0 },
		{ SIGUSR1, "SIGUSR1", 1 },
		{ SIGUSR2, "SIGUSR2", 1 },
		{ SIGPROF, "SIGPROF", 1 },
		{ SIGVTALRM, "SIGVTALRM", 1 },
#ifdef VDE_LINUX
		{ SIGPOLL, "SIGPOLL", 1 },
		{ SIGSTKFLT, "SIGSTKFLT", 1 },
		{ SIGIO, "SIGIO", 1 },
		{ SIGPWR, "SIGPWR", 1 },
		{ SIGUNUSED, "SIGUNUSED", 1 },
#endif
#ifdef VDE_DARWIN
		{ SIGXCPU, "SIGXCPU", 1 },
		{ SIGXFSZ, "SIGXFSZ", 1 },
#endif
		{ 0, NULL, 0 }
	};

	int i;
	for(i = 0; signals[i].sig != 0; i++)
		if(signal(signals[i].sig,
					signals[i].ignore ? SIG_IGN : sig_handler) < 0)
			printlog(LOG_ERR,"Setting handler for %s: %s", signals[i].name,
					strerror(errno));
}

void set_switchmac()
{
	struct timeval v;
	long val;
	register int i;
	gettimeofday(&v,NULL);
	srand48(v.tv_sec ^ v.tv_usec ^ getpid());
	for(i=0,val=lrand48();i<4; i++,val>>=8)
		switchmac[i+2]=val;
	switchmac[0]=0;
	switchmac[1]=0xff;
}

static void start_modules(void);

int main(int argc, char **argv)
{
	set_switchmac();
	start_modules();
	parse_args(argc,argv);
	atexit(cleanup);
	setsighandlers();
	hash_init(hash_size);
#ifdef FSTP
	fst_init(numports);
#endif
	port_init(numports);
	init_mods();
	loadrcfile();
	qtimer_init();
	main_loop();
	return 0;
}

/* modules: module references are only here! */
static void start_modules(void)
{
	void start_consmgmt(void);
	void start_datasock(void);
	void start_tuntap(void);
	start_consmgmt();
	start_datasock();
	start_tuntap();
}
